### PR TITLE
Add osu! mod "Approach Different"

### PR DESCRIPTION
**Pull request by peppy**
Taking over from https://github.com/ppy/osu/pull/8284/ as I can't push to the branch there. Credit to @BlauFx for the idea / initial implementation.

Have renamed and tidied up the easing naming and mapping logic.

https://user-images.githubusercontent.com/191335/122597130-e7149880-d0a5-11eb-9853-458a3f3f6fa3.mp4

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace osu.Game.Rulesets.Osu.Mods
+{
+    internal class OsuModDifferentApproach : Mod, IApplicableToDrawableHitObjects
+    {
+        public override string Name => "Approach Different";
+        public override string Acronym => "AD";
+        public override string Description => "Never trust the approach circles...";
+        public override double ScoreMultiplier => 1;
+        public override IconUsage? Icon { get; } = FontAwesome.Regular.Circle;
```
Comment (jvyden) [inline osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs:23]:
```suggestion
        public override IconUsage? Icon { get; } = FontAwesome.Regular.QuestionCircle;
```

File: osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Osu.Objects.Drawables;
+
+namespace osu.Game.Rulesets.Osu.Mods
+{
+    internal class OsuModDifferentApproach : Mod, IApplicableToDrawableHitObjects
```
Comment (jvyden) [inline osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs:17]: This should be changed to `OsuModApproachDifferent`.

File: osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs:23]: No thanks

File: osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        public override string Name => "Approach Different";
+        public override string Acronym => "AD";
+        public override string Description => "Never trust the approach circles...";
+        public override double ScoreMultiplier => 1;
+        public override IconUsage? Icon { get; } = FontAwesome.Regular.Circle;
+
+        [SettingSource("Easing", "Change the animation curve of the approach circles.", 0)]
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs:25]: This reads worse.

[the remark this answers is not in the document]
Comment (jvyden) [reply osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs]: > This reads worse.

Fair enough for the description, but the term `Easing` doesn't really sit right in my eyes, at least from a player's point of view.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs]: Oh right. That should be changed to "style".

[the remark this answers is not in the document]
Comment (jvyden) [reply osu.Game.Rulesets.Osu/Mods/OsuModDifferentApproach.cs]: Ah, that's the word I was looking for. I'll update the suggestion.

File: osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+{
+    public class OsuModApproachDifferent : Mod, IApplicableToDrawableHitObjects
+    {
+        public override string Name => "Approach Different";
+        public override string Acronym => "AD";
+        public override string Description => "Never trust the approach circles...";
+        public override double ScoreMultiplier => 1;
+        public override IconUsage? Icon { get; } = FontAwesome.Regular.Circle;
+
+        [SettingSource("Easing", "Change the animation curve of the approach circles.", 0)]
```
Comment (jvyden) [inline osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:25]:
```suggestion
        [SettingSource("Style", "Change the animation curve of the approach circles.", 0)]
```

File: osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        {
+            drawables.ForEach(drawable =>
+            {
+                drawable.ApplyCustomUpdateState += (drawableHitObj, state) =>
+                {
+                    if (!(drawableHitObj is DrawableHitCircle hitCircle)) return;
+
+                    var obj = hitCircle.HitObject;
+
+                    hitCircle.BeginAbsoluteSequence(obj.StartTime - obj.TimePreempt, true);
```
Comment (ekrctb) [inline osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:48]: Shouldn't `BeginAbsoluteSequence` be use with `using`?

File: osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                    hitCircle.ApproachCircle.FadeIn(Math.Min(obj.TimeFadeIn, obj.TimePreempt));
+
+                    hitCircle.ApproachCircle.ScaleTo(1f, obj.TimePreempt, getEasing(Style.Value));
+
+                    hitCircle.ApproachCircle.Expire(true);
+                };
+            });
+        }
+
+        private Easing getEasing(AnimationStyle approachEasing)
```
Comment (ekrctb) [inline osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:60]: Nit: the parameter name is still using "easing" while the type name is renamed to "style".

File: osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                    return Easing.InOutCubic;
+
+                case AnimationStyle.InOut2:
+                    return Easing.InOutQuint;
+            }
+        }
+
+        public enum AnimationStyle
+        {
+            Default,
```
Comment (ekrctb) [inline osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:98]: I wonder the mod should use an animation style other than the `Easing.None` by default.
A user will see no difference to the gameplay when the mod is enabled until the user finds the customization menu.

**Comment by ekrctb** [comment]
(Not specific to this PR)
In 16∶9 resolution, the mod button is not clickable when the customization menu is opening because the menu is overlapping the mod button.

File: osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs (hunk 7 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs:98]: Fair call. I'm not sure if we even want default on here..

**Review by ekrctb — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]

**Comment by BlauFx** [comment]
Thanks @peppy for picking this one up, I completely forgot about it. I'm glad to see, that this mod made it's way into osu!.
Thank you

**Pull request merged**